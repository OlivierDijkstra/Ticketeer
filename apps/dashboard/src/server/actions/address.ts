'use server';

import type { Address } from '@repo/lib';

import { fetchWithAuth } from '@/lib/fetch';

export async function getAddressAction({ address_id }: { address_id: string }) {
  return await fetchWithAuth<Address>(`api/addresses/${address_id}`);
}

export async function updateAddressAction({
  address_id,
  data,
}: {
  address_id: string;
  data: Partial<Address>;
}) {
  return await fetchWithAuth<Address>(`api/addresses/${address_id}`, {
    method: 'PUT',
    body: data,
  });
}