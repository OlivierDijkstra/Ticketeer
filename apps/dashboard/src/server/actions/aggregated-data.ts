'use server';

import { createUrl } from '@repo/lib';

import type { DateRanges, ResultSet } from '@/components/charts/lib';
import { fetchWithAuth } from '@/lib/fetch';

export type AggregationType = 'count' | 'sum' | 'avg' | 'min' | 'max';
export type Granularity = 'hour' | 'day' | 'week' | 'month' | 'year';

export type AggregatedDataConfig = {
  modelType: string;
  aggregationType: AggregationType;
  granularity: Granularity;
  dateRange: DateRanges | [Date, Date];
};

export async function fetchAggregatedData({
  modelType,
  aggregationType,
  granularity,
  dateRange,
}: AggregatedDataConfig) {
  const date_range =
    typeof dateRange === 'string' ? dateRange : JSON.stringify(dateRange);

  const url = createUrl(`api/aggregations`, {
    model_type: modelType,
    aggregation_type: aggregationType,
    granularity,
    date_range,
  });

  const response = await fetchWithAuth<{
    data: ResultSet;
  }>(url);

  return response.data;
}
