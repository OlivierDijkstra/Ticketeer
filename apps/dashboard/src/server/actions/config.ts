'use server';

export async function getConfig() {
  const result = {
    APP_LOCALE: process.env.APP_LOCALE || 'en-US',
    APP_CURRENCY: process.env.APP_CURRENCY || 'USD',
    APP_NAME: process.env.APP_NAME || 'Dashboard',
  };

  return result;
}
