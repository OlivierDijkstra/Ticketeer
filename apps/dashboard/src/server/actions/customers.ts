'use server';

import { fetchWithAuth } from '@/lib/fetch';
import { createUrl } from '@/lib/utils';
import type { Customer, PaginatedResponse } from '@/types/api';

export async function getCustomersAction({
  search,
  page,
  sorting,
  show_id,
}: {
  search?: string;
  page?: string;
  sorting?: { id: string; desc: boolean };
  show_id?: number;
}) {
  const url = createUrl('api/customers', {
    search,
    page: page || '1',
    sort: JSON.stringify(sorting),
    show_id: show_id,
  });

  return await fetchWithAuth<PaginatedResponse<Customer>>(url, {
    next: {
      tags: ['customers'],
    },
  });
}

export async function getCustomerAction({
  customer_id,
}: {
  customer_id: string;
}) {
  return await fetchWithAuth<Customer>(`api/customers/${customer_id}`, {
    next: {
      tags: ['customers'],
    },
  });
}
