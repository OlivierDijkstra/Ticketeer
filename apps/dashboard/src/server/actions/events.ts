'use server';

import { fetchWithAuth } from '@/lib/fetch';
import { createUrl } from '@/lib/utils';
import type { CreateEvent, Event, PaginatedResponse } from '@/types/api';

export async function getEventsAction({
  page,
}: {
  page?: string;
} = {}) {
  const url = createUrl('api/events', {
    page: page || 1,
  });

  return await fetchWithAuth<PaginatedResponse<Event>>(url, {
    next: {
      tags: ['events'],
    },
  });
}

export async function getEventAction({ event_slug }: { event_slug: string }) {
  return await fetchWithAuth<Event>(`api/events/${event_slug}`);
}

export async function createEventAction({ data }: { data: CreateEvent }) {
  return await fetchWithAuth<Event>('api/events', {
    method: 'POST',
    body: data,
  });
}

export async function updateEventAction({
  event_slug,
  data,
}: {
  event_slug: string;
  data: Partial<CreateEvent>;
}) {
  return await fetchWithAuth<Event>(`api/events/${event_slug}`, {
    method: 'PUT',
    body: data,
  });
}

export async function setEventFeaturedAction({
  event_slug,
}: {
  event_slug: string;
}) {
  return await fetchWithAuth<Event>(`api/events/${event_slug}`, {
    method: 'PUT',
    body: {
      featured: true,
    },
  });
}

export async function addEventMediaAction({
  event_slug,
  data,
}: {
  event_slug: string;
  data: FormData;
}) {
  return await fetchWithAuth<Event>(`api/events/${event_slug}/media`, {
    method: 'POST',
    body: data,
  });
}

export async function deleteEventMediaAction({
  event_slug,
  media_id,
}: {
  event_slug: string;
  media_id: number;
}) {
  return await fetchWithAuth<Event>(
    `api/events/${event_slug}/media/${media_id}`,
    {
      method: 'DELETE',
    }
  );
}

export async function setEventCoverAction({
  event_slug,
  media_id,
}: {
  event_slug: string;
  media_id: number;
}) {
  return await fetchWithAuth<Event>(
    `api/events/${event_slug}/media/${media_id}/cover`,
    {
      method: 'PUT',
    }
  );
}
