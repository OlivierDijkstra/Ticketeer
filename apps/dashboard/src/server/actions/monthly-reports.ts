'use server';

import type { MonthlyReport } from '@repo/lib';

import { fetchWithAuth } from '@/lib/fetch';

export async function getMonthlyReports() {
  return await fetchWithAuth<MonthlyReport[]>('api/monthly-reports', {
    method: 'GET',
    next: {
      tags: ['monthly-reports'],
    },
  });
}
