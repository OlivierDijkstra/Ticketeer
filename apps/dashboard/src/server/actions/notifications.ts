'use server';

import { fetchWithAuth } from '@/lib/fetch';

export async function notifyTickets({ order_id }: { order_id: string }) {
  await fetchWithAuth('api/notifications/tickets', {
    method: 'POST',
    body: {
      order_id,
    },
  });
}
