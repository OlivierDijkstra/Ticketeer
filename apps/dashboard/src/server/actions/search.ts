'use server';

import { fetchWithAuth } from '@/lib/fetch';
import { createUrl } from '@/lib/utils';
import type { Customer, Event, Product, Show } from '@/types/api';

export async function searchAction({
  query,
  filter,
}: {
  query: string;
  filter?: string;
}) {
  const url = createUrl('api/search', { query, filter });

  return await fetchWithAuth<{
    events: Event[];
    products: Product[];
    shows: Show[];
    customers: Customer[];
  }>(url, {
    next: {
      tags: ['search'],
    },
  });
}
