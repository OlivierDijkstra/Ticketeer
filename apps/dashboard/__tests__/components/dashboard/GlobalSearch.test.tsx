import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import type { useRouter } from 'next/navigation';
import type { Mock } from 'vitest';
import { describe, expect, test, vi } from 'vitest';

import GlobalSearch from '@/components/GlobalSearch';
import { searchAction } from '@/server/actions/search';

const routerMock = {
  push: vi.fn(),
};

vi.mock('@/server/actions/search', () => ({
  searchAction: vi.fn(),
}));

vi.mock('next/navigation', () => ({
  useRouter: () => routerMock as unknown as ReturnType<typeof useRouter>,
}));

describe('GlobalSearch', () => {
  test('it opens and closes the command dialog', async () => {
    render(<GlobalSearch />);

    const user = userEvent.setup();

    await user.keyboard('{Meta>}{k}{/Meta}');

    const input = screen.getByPlaceholderText('Type a command or search...');
    expect(input).toBeInTheDocument();

    await user.keyboard('{Escape}');
    expect(input).not.toBeInTheDocument();
  });

  test('it searches and displays results correctly', async () => {
    render(<GlobalSearch />);

    const user = userEvent.setup();

    await user.keyboard('{Meta>}{k}{/Meta}');

    const input = screen.getByPlaceholderText('Type a command or search...');
    await user.type(input, 'event');

    const eventsGroup = screen.getByText('Suggestions');
    expect(eventsGroup).toBeInTheDocument();

    const events = screen.getByRole('option', {
      name: /events/i,
    });
    expect(events).toBeInTheDocument();
  });

  test('it navigates correctly on selecting a search result', async () => {
    render(<GlobalSearch />);
    const user = userEvent.setup();

    // Simulate key press to open the dialog
    await user.keyboard('{Meta>}{k}{/Meta}');

    const input = screen.getByPlaceholderText('Type a command or search...');
    await user.type(input, 'event');

    const menuItem = screen.getByRole('option', {
      name: /events/i,
    });

    await user.click(menuItem);

    await waitFor(() => {
      expect(routerMock.push).toHaveBeenCalledWith('/dashboard/events');
    });
  });

  test('it clears the search input when the dialog is closed and reopened', async () => {
    render(<GlobalSearch />);
    const user = userEvent.setup();

    await user.keyboard('{Meta>}{k}{/Meta}');
    const input = screen.getByPlaceholderText('Type a command or search...');
    await user.type(input, 'event');

    await user.keyboard('{Escape}');
    expect(input).not.toBeInTheDocument();

    await user.keyboard('{Meta>}{k}{/Meta}');
    const newInput = screen.getByPlaceholderText('Type a command or search...');
    expect(newInput).toHaveValue('');
  });

  test('it displays different categories in the search results', async () => {
    (searchAction as Mock).mockResolvedValue({
      events: [
        { id: '1', name: 'Event 1' },
        { id: '2', name: 'Event 2' },
      ],
      products: [
        { id: '1', name: 'Product 1' },
        { id: '2', name: 'Product 2' },
      ],
      shows: [
        {
          id: '1',
          name: 'Show 1',
          start: '2022-01-01T00:00:00Z',
          end: '2022-01-01T00:00:00Z',
          guests: [],
        },
        {
          id: '2',
          name: 'Show 2',
          start: '2022-01-01T00:00:00Z',
          end: '2022-01-01T00:00:00Z',
          guests: [],
        },
      ],
    });

    render(<GlobalSearch />);
    const user = userEvent.setup();

    await user.keyboard('{Meta>}{k}{/Meta}');

    const input = screen.getByPlaceholderText('Type a command or search...');
    await user.type(input, 'show');

    await waitFor(() => {
      expect(searchAction).toHaveBeenCalledWith({ query: 'show' });

      const eventsGroup = screen.getByText('Events');
      const productsGroup = screen.getByText('Products');
      const showsGroup = screen.getByText('Shows');

      expect(productsGroup).toBeInTheDocument();
      expect(eventsGroup).toBeInTheDocument();
      expect(showsGroup).toBeInTheDocument();
    });
  });
});
