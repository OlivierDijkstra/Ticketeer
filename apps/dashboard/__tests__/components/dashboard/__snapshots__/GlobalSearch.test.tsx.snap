// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GlobalSearch > it renders correctly 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center rounded-md font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 group whitespace-nowrap text-sm text-muted-foreground"
  >
    <svg
      class="lucide lucide-search mr-2 size-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="11"
        cy="11"
        r="8"
      />
      <path
        d="m21 21-4.3-4.3"
      />
    </svg>
    Search...
    <span
      class="ml-8"
    >
      <kbd
        class="pointer-events-none inline-flex h-5 select-none items-center gap-1 rounded border bg-muted px-1.5 font-mono text-[10px] font-medium text-muted-foreground opacity-100 transition-colors group-hover:bg-muted-foreground/30"
      >
        <span
          class="text-xs"
        >
          âŒ˜
        </span>
        K
      </kbd>
    </span>
  </button>
</div>
`;
